hack/flow-test/main.go

```diff
@@ -0,0 +1,315 @@
+// shipperizer in ~/shipperizer/hydra on feat_dev_grants_2x λ ./hydra create client --endpoint http://127.0.0.1:4445 --name auth --grant-type authorization_code,refresh_token,urn:ietf:params:oauth:grant-type:device_code --response-type code,id_token --scope openid,offline --redirect-uri http://localhost:8000/api/ready
+// CLIENT ID     efc05555-e960-4aa2-bc0d-291144e15963
+// CLIENT SECRET   cjMbxFlwS1BHXEyVpR-3JLoTPN
+// GRANT TYPES     authorization_code, refresh_token, urn:ietf:params:oauth:grant-type:device_code
+// RESPONSE TYPES  code, id_token
+// SCOPE           openid offline
+// AUDIENCE
+// REDIRECT URIS   http://localhost:8000/api/ready
+
+type ProviderType int
+
+const (
+	Hydra ProviderType = iota
+	Github
+)
+
+// EnvSpec is the basic environment configuration setup needed for the app to start
+type EnvSpec struct {
+	OAuthClientID     string       `envconfig:"oauth_client_id"`
+	OAuthClientSecret string       `envconfig:"oauth_client_secret"`
+	CallbackURI       string       `envconfig:"callback_uri" default:"http://localhost:8000/api/ready"`
+	Scopes            []string     `envconfig:"scopes" default:"openid,offline"`
+	Provider          ProviderType `envconfig:"provider" default:"0"`
+	AuthURL           string       `envconfig:"auth_url" default:"http://localhost:4444/oauth2/auth"`
+	TokenURL          string       `envconfig:"token_url" default:"http://localhost:4444/oauth2/token"`
+	DeviceAuthURL     string       `envconfig:"device_auth_url" default:"http://localhost:4444/oauth2/device/auth"`
+	HydraAdminApiURL  string       `envconfig:"hydra_admin_api_url" default:"http://localhost:4445"`
+}
+
+func registerHydraClient(hydraAdminUrl string) string {
+	configuration := client.NewConfiguration()
+	configuration.Servers = []client.ServerConfiguration{
+		{
+			URL: hydraAdminUrl,
+		},
+	}
+
+	configuration.HTTPClient = &http.Client{Transport: otelhttp.NewTransport(http.DefaultTransport)}
+
+	c := client.NewAPIClient(configuration)
+
+	oauthClient := client.NewOAuth2Client()
+	oauthClient.SetGrantTypes([]string{"authorization_code", "refresh_token", "urn:ietf:params:oauth:grant-type:device_code"})
+	oauthClient.SetScope("openid profile offline email")
+	oauthClient.SetTokenEndpointAuthMethod("none")
+
+	cc, _, err := c.OAuth2Api.CreateOAuth2Client(context.Background()).OAuth2Client(*oauthClient).Execute()
+	if err != nil {
+		panic("Failed to create oauth2 client " + err.Error())
+	}
+	return *cc.ClientId
+}
+
+func deviceFlow(specs *EnvSpec, logger *zap.SugaredLogger) {
+	config := new(oauth2.Config)
+	config.ClientID = specs.OAuthClientID
+	config.ClientSecret = specs.OAuthClientSecret
+	config.Scopes = specs.Scopes
+
+	switch specs.Provider {
+	case Github:
+		config.Endpoint = github.Endpoint
+	case Hydra:
+		config.Endpoint = oauth2.Endpoint{
+			AuthURL:       specs.AuthURL,
+			TokenURL:      specs.TokenURL,
+			DeviceAuthURL: specs.DeviceAuthURL,
+			AuthStyle:     oauth2.AuthStyleInHeader,
```

Comment:
This is needed because otherwise the go client will try to hit the token endpoint again when it gets an `authorization_pending` response. The reason for this is that if no client authn method is specified, the client will try to hit the token endpoint with another auth method if it gets a token endpoint error. The problem is that it does not parse the error code to validate that the error was caused by a `client_authn_failed`.